### ISSUE TITLE
Uncaught SyntaxError: Unexpected end of JSON input

### ISSUE THREAD

**Issue by mkborregaard** [AtomLinter/linter-julia#38]
Happens when I open a Julia project in Atom.

**Atom**: 1.15.0 x64
**Electron**: 1.3.13
**OS**: Mac OS X 10.12.3
**Thrown From**: [linter-julia](https://github.com/TeroFrondelius/linter-julia) package 0.7.1

### Stack Trace

Uncaught SyntaxError: Unexpected end of JSON input

```
At file:///Applications/Atom.app/Contents/Resources/app.asar/static/index.html:1

SyntaxError: Unexpected end of JSON input
    at Object.parse (native)
    at /packages/linter-julia/lib/linter-julia.coffee:93:26)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at Pipe._handle.close [as _onclose] (net.js:493:12)
```

### Commands

```
```

### Non-Core Packages

```
2-dark-syntax 1.1.0 
actual-solarized-syntax 0.1.0 
atom-dark-fusion-syntax 2.0.3 
atom-material-syntax 1.0.2 
atom-material-ui 1.3.9 
atom-solarized-dark-ui 0.28.0 
autocomplete-bibtex 0.7.0 
autoupdate-packages 1.3.1 
color-picker 2.2.5 
compare-files 0.8.0 
export-html 3.1.0 
focus-dark 1.1.0 
idle-theme 0.1.0 
idledark 0.3.1 
idlefingers 0.2.0 
indent-detective 0.1.0 
ink 0.6.5 
julia-client 0.5.12 
language-julia 0.10.0 
language-matlab-only 0.2.0 
language-r 0.4.1 
latex-completions 0.3.2 
linter 1.11.23 
linter-julia 0.7.1 
one-dark-shade-ui 0.1.0 
polaris-syntax 0.1.0 
python-jedi 0.3.8 
python-tools 0.6.9 
razor-atom-syntax-theme 2.2.3 
tomorrow-night-bright-razor-atom-syntax-theme 1.0.0 
solarized-dark-ui 0.3.2 
solarized-light-syntax-modified 0.6.2 
solarized-light-ui 0.1.0 
solarized-seti-ui 0.8.6 
solo-one-dark-ui 0.1.1 
todo-show 1.11.0 
tomorrow-night-bright-razor-atom-syntax-theme 1.0.0 
tomorrow-night-bright-syntax 0.1.0 
tool-bar 1.1.0 
twilight 0.3.0 
twilight-syntax 0.7.0 
twilight-textmate-syntax 0.1.19 
uber-juno 0.1.1 
wordcount 2.10.4 
```

**Comment by TeroFrondelius** [comment]
Have you run `Pkg.checkout("Lint")`?  Also restarting Atom (Ctrl + Shift + F5) may help. Also Ctrl + Shift + I, opens developer tools and from there the output is interesting: namely is the lintserver running.

**Comment by dpo** [comment]
I have the same problem. I ran `Pkg.checkout("Lint")` but the result is the same. The dev tools say 
```
linter-julia.coffee:72 connection ended with error ArgumentError("stream is closed or unusable")
```

**Comment by xukai92** [comment]
I have the same error. I guess it's because the recent update of the Lint.jl package?

**Comment by razhangwei** [comment]
I had the same error. 

[Enter steps to reproduce:]

1. ...
2. ...

**Atom**: 1.15.0 x64
**Electron**: 1.3.13
**OS**: Mac OS X 10.11.6
**Thrown From**: [linter-julia](https://github.com/TeroFrondelius/linter-julia) package 0.7.1


### Stack Trace

Uncaught SyntaxError: Unexpected end of JSON input

```
At file:///Applications/Atom.app/Contents/Resources/app.asar/static/index.html:1

SyntaxError: Unexpected end of JSON input
    at Object.parse (native)
    at /packages/linter-julia/lib/linter-julia.coffee:93:26)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at Pipe._handle.close [as _onclose] (net.js:493:12)
```

### Commands

```
     -1:46.6.0 core:backspace (input.hidden-input)
  5x -1:45.4.0 core:save (input.hidden-input)
     -1:35.2.0 core:backspace (input.hidden-input)
     -1:34.8.0 core:save (input.hidden-input)
```

### Non-Core Packages

```
atom-beautify 0.29.17 
atom-latex 0.7.1 
autocomplete-python 1.8.60 
busy-signal 1.3.0 
docblockr 0.9.1 
Hydrogen 1.6.0 
intentions 1.1.2 
language-julia 0.10.0 
language-latex 1.0.0 
last-cursor-position 0.9.0 
latex-completions 0.3.2 
linter 2.1.0 
linter-julia 0.7.1 
linter-pylint 1.2.2 
linter-ui-default 1.1.0 
minimap 4.26.8 
project-manager 3.3.3 
Remote-FTP 0.10.7 
sync-settings 0.8.1 
termination 0.5.0 
```

**Comment by mkborregaard** [comment]
The number of errors thrown by this package went through the roof. I will reinstall in a few months. Keep up the good work, it looks like it will become extremely useful with time!

**Comment by TeroFrondelius** [comment]
I will look into the issue today.

**Comment by mkborregaard** [comment]
It looks like it may actually the linter-ui-default package that is responsible for much of the trouble.

**Comment by TeroFrondelius** [comment]
This is a real problem. Here is the current work around:
```bash
cd ~/.julia/v0.5/Lint
git checkout 1dfcfd4
cd ~/.atom/packages/linter-julia
git checkout v0.5.5 
```
Above is the solution to go back before the JSON change is implemented.

**Comment by mkborregaard** [comment]
Does that also work if you've installed linter-julia inside Atom (rather than cloning the repo and `apm`)? I get 
```
git checkout v0.5.5
fatal: Not a git repository (or any of the parent directories): .git
```

**Comment by TeroFrondelius** [comment]
Here is the guideline for Atom packages: https://discuss.atom.io/t/how-do-you-download-a-specific-version-of-a-package/11124/4

```bash
apm install <package_name>@<package_version>
```
Thus it should be: 
```bash
apm install linter-julia@v0.5.5
```

**Comment by mkborregaard** [comment]
```
zn11265:linter-julia michael$ apm install linter-julia@v0.5.5
Installing linter-julia@v0.5.5 to /Users/michael/.atom/packages ✗
Package version: v0.5.5 not found
```

**Comment by TeroFrondelius** [comment]
@mkborregaard can you try one more: 
```bash
apm install linter-julia@0.5.5
```

**Comment by mkborregaard** [comment]
Now I get :-)
```
Installing linter-julia@0.5.5 to /Users/michael/.atom/packages ✓
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
zn11265:linter-julia michael$
```

**Comment by TeroFrondelius** [comment]
I have now tested with both linux and windows computers. linter-julia works after `Pkg.free("Lint")` and update to latest linter-julia version 0.7.1. On linux my mistake was not to run the `Pkg.free("Lint")` properly, which meant that the requirements was not updated, my problem was caused an old version of Compat.jl. Windows I had to set the path to `julia.exe`, because the juno default is still julia version 0.4. 

**After updates on both platforms I had to restart Atom at least one time by pressing Ctrl + Shift + F5, to get the linter-julia working** (see #40) 

I will develop some tests (specs) for linter-julia and try to improve the error messages as well as the logic of auto-installer or possible auto-updater. Although I am little worried that too much automation without user action is unwanted feature some developers, who want to control the updates manually. 

If the problem persists, please post `Pkg.status()` output.

**Comment by xukai92** [comment]
@TeroFrondelius Thank you for the information. I just tried it on my Ubuntu machine, which turned out to work now! I will later try it on macOS and update the info.

**Comment by TeroFrondelius** [comment]
@xukai92 thanks a lot for testing.

**Comment by dpsanders** [comment]
I can confirm that this seems to fix the problem on MacOS, thanks!

**Comment by mkborregaard** [comment]
Mine are consistent, but seem to be from linter-ui-default so I commented on an issue there.

**Comment by mkborregaard** [comment]
I can probably close this now? Or leave open so people can find the fix?

**Comment by xukai92** [comment]
@dpsanders I still experienced this problem on my macOS. Can you plz tell me your linter-julia version and Lint.jl version please?

**Comment by dpsanders** [comment]
Correction, I still have an error.

**Comment by dpsanders** [comment]
```/Users/dpsanders/.atom/packages/linter/lib/linter-registry.js:177 [Linter] Error running linter-julia Error: write EPIPE(…)(anonymous function) @ /Users/dpsanders/.atom/packages/linter/lib/linter-registry.js:177

index.html:1 Uncaught SyntaxError: Unexpected end of JSON input

/Users/dpsanders/.atom/packages/linter-julia/lib/linter-julia.coffee:72 WARNING: could not import Base.iszero into Main

/Users/dpsanders/.atom/packages/linter-julia/lib/linter-julia.coffee:72 connection ended with error EOFError()
```

**Comment by dpsanders** [comment]
```
SyntaxError: Unexpected end of JSON input
    at Object.parse (native)
    at Socket.<anonymous> (/Users/dpsanders/.atom/packages/linter-julia/lib/linter-julia.coffee:93:26)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at Pipe._handle.close [as _onclose] (net.js:493:12)
```

**Comment by dpsanders** [comment]
Versions:
- Lint.jl 0.3
- linter 2.1.0
- linter-julia 0.7.1
- linter-ui-default 1.2.1

**Comment by TeroFrondelius** [comment]
My versions (Windows):
* Lint.jl 0.3.0
* linter 1.11.21
* linter-julia 0.7.1
* linter-ui-default not installed

This would indicate to me that maybe the reason to this bug is the change to linter 2.1.0.

**Comment by xukai92** [comment]
My fail on macOS used:

- Lint 0.3.0
- linter 2.1.0
- linter-julia 0.7.1
- linter-ui-default 1.2.1

**Comment by dpsanders** [comment]
FWIW, I'm still seeing the error after updating to `Lint.jl` v0.4.

**Comment by TeroFrondelius** [comment]
If someone could try to make clean installation of 0.7.2, it would be really appreciated. You probably need to uninstall linter and linter-ui-default before updating linter-julia.

**Comment by cynddl** [comment]
I did a clean install of linter-julia, linter, and linter-ui-default. The problem seems to still be present. I currently have:

- Lint 0.3.0
- linter 1.11.21
- linter-julia 0.7.2
- linter-ui-default 1.2.1

```
SyntaxError: Unexpected end of JSON input
    at Socket.<anonymous> (/home/cynddl/.atom/packages/linter-julia/lib/linter-julia.coffee:93:26)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at Pipe._handle.close [as _onclose] (net.js:493:12)
```

**Comment by TeroFrondelius** [comment]
@cynddl please uninstall linter-ui-default (if you won't need it for other linter), I think that is causing the problem. linter-ui-default is not yet needed with linter-julia. 

I tried clean install on my windows machine, with these steps:
1. update Lint.jl v"0.4.0"
2. run `Pkg.test("Lint")` (in order to pre-compile Lint.jl) 
3. uninstall linter, linter-julia
4. set correct julia path to juno (julia-client settings)
5. install linter-julia from settings, which asked to install linter 1.11.21

I also noticed that my autoinstaller #30 isn't working.

**Comment by cynddl** [comment]
linter-ui-default was not causing the issue, as I tried with and without. I installed Lint.jl v0.4.0 and did a clean install of linter-julia and linter (1.11.21, from settings). The error message is still there.

**Comment by TeroFrondelius** [comment]
@cynddl can you see the lintserver running in console, something like `Server running on port/pipe \\.\pipe\lintserver6504nKjRJXcaGX2Ssock ...` if the error comes after this message, then press Ctrl + Shift + F5 to restart Atom (or should it down and restart). That problem is issue #40, although not so well explained.

**Comment by cynddl** [comment]
Yes, I get both:

    Server running on port/pipe /tmp/lintserver22650PAngS3vuAHCIsock ...

and:

    Uncaught SyntaxError: Unexpected end of JSON input

Also, it's followed by this message that didn't seem to be there before:

    connection ended with error ArgumentError("stream is closed or unusable")

After restarting Atom (with F5), the errors messages are still happening (altough the *connected ended* comes before the *SyntaxError*)

**Comment by TeroFrondelius** [comment]
@cynddl can you try to change linter settings: Lint As You Type Interval for something bigger than default 300, for example 1000. BTW Ctrl + s starts the linting, sometimes when I did debugging I pressed Ctrl + s multiple times to get the server - client to sync. I am not sure but I have the feeling that error message is because Lint.jl lintserver expects reading, but linter-julia is trying to write to the server (or vice versa). So far I have always get rid of this error when I get it working once.

**Comment by fufjvnvnf** [comment]
I'm currently experiencing this exception on atoms #46

**Comment by aminya** [comment]
Could you try the latest version? Does this error still exit?
https://atom.io/packages/linter-julia

**Comment by TeroFrondelius** [comment]
Most probably Lint.jl precombillation causes this issue. The lintserver is not up and running yet. Pressing Ctrl+s (saving the file) few times may resolve it or restarting Atom.

**Comment by TeroFrondelius** [comment]
Sorry for the noise. This was a really old issue.

**Comment by aminya** [comment]
I do not have any issues. Are on the master repository of Lint.jl?

**Comment by aminya** [comment]
Opening a project (a folder) gives the error.

**Comment by aminya** [comment]
Fixed in #119
